### PR TITLE
Fix skipping messages logic + performance improvements

### PR DESCRIPTION
**Pull request by nuno-vieira**
### 🎯 Goal
Fixes the logic around skipping messages when receiving messages and scrolling at the top.
It also improves the efficiency of skipping messages by transforming the `LazyCachedMapCollection` to `Array` only when there are message list UI changes.

### 🎨 Showcase
To be filled when merged to develop

### 🧪 Manual Testing Notes
Run the performance load script while scrolling the message list to the top. The message list should not jump + performance should be good.

### ☑️ Contributor Checklist
To be filled when merged to develop

### CODE REVIEW

File: Sources/StreamChat/Utils/LazyCachedMapCollection.swift [hunk 1]
```diff
@@ -70,6 +70,10 @@ public struct LazyCachedMapCollection<Element>: RandomAccessCollection {
             return value
         }
     }
+
+    public func append(_ element: Element) {
```
Comment (martinmitrevski) [inline Sources/StreamChat/Utils/LazyCachedMapCollection.swift:74]: Where do you use this one?

**Review by martinmitrevski — COMMENTED** [review]
looks good, just one question. @testableapple, please start with the QA as well.

File: Sources/StreamChat/Utils/LazyCachedMapCollection.swift (hunk 1 above)
Comment (nuno-vieira) [reply Sources/StreamChat/Utils/LazyCachedMapCollection.swift:74]: It is not shown in the changes, but we are appending the Root of the thread in the `getReplies()` function at the end. Previously we were using the `Array`. Now we need for `LazyCachedMapCollection` to support this as well.

File: Sources/StreamChatUI/ChatMessageList/ChatMessageListVC.swift [hunk 2]
```diff
@@ -1052,31 +1054,28 @@ private extension ChatMessageListVC {
...
     }
 
-    func addSkippedMessagesIfNeeded(with changes: [ListChange<ChatMessage>], newestChange: ListChange<ChatMessage>?) {
+    func shouldSkipMessages(with changes: [ListChange<ChatMessage>], newestChange: ListChange<ChatMessage>?) -> Bool {
         let insertions = changes.filter(\.isInsertion)
         let isNewestChangeInsertion = newestChange?.isInsertion == true
         let isNewestChangeNotByCurrentUser = newestChange?.item.isSentByCurrentUser == false
         let isNewestChangeNotVisible = !listView.isLastCellFullyVisible && !listView.previousMessagesSnapshot.isEmpty
-        let hasMultipleInsertions = insertions.count > 1
+        let isNewPageLoading = insertions.count == changes.count
```
Comment (polqf) [inline Sources/StreamChatUI/ChatMessageList/ChatMessageListVC.swift:1062]:
```suggestion
        let isLoadingNewPage = insertions.count == changes.count
```

File: Sources/StreamChatUI/ChatThread/ChatThreadVC.swift [hunk 3]
```diff
@@ -125,9 +125,6 @@ open class ChatThreadVC: _ViewController,
             queueAudioPlayer.dataSource = self
         }
 
-        // Set the initial data
-        messages = getReplies(from: messageController)
```
Comment (polqf) [inline Sources/StreamChatUI/ChatThread/ChatThreadVC.swift:129]: Not setting initial data anymore?

Comment (nuno-vieira) [reply Sources/StreamChatUI/ChatThread/ChatThreadVC.swift:129]: This was reverted actually 👍

**Review by martinmitrevski — APPROVED** [review]

**Pull request merged**